design/specs/specs-null-values.md

```diff
@@ -0,0 +1,124 @@
+description: 
+published: true
+date: 2021-11-25T09:00:00.201Z
+tags: 
+editor: markdown
+dateCreated: 2021-11-25T09:00:00.201Z
+---
+
+## Context
+
+NULL values are used to indicate that a value does not exist in a database. It is important to make it clear to users that NULL is not the same as blank or zero. Ideally, we want users to think of NULL values as placeholders for unknown or missing values that are applicable.
+
+## Prototype
+
+### Link
+
+[Prototype](https://mathesar-prototype.netlify.app/?activeTable=2)
+
+### Videos
+
+- Scenario 1: [Video](https://www.loom.com/share/527224982d534a2db082359360147ad4)
+- Scenario 2: [Video](https://www.loom.com/share/7de3c90af3d74ac195397c643f318867)
+- Scenario 3: [Video](https://www.loom.com/share/7ea4ca9fe3e14acdbffac5ea8e77e760)
+
+## User Experience
+
+### Considerations for handling NULL Values
+
+When presented in the form of tabular data, NULL values should be visually distinct so that the user understands that they are not the same as other string types. The style should attempt to communicate the absence of a value, similar to how placeholders are used in text input fields.
+As for the process of transforming values to and from NULL, it will be important to have a consistent way of doing this across different data types.
+
+## Scenarios
+
+### Scenario 1: User identifies a NULL Value
+
+#### Scenario 1a: The user identifies a NULL value in a table
+
+#### Steps for 1a
+
+1. The user opens a table that contains NULL values
+2. The user identifies a cell that contains a NULL value
+    - The cell contains the word NULL in all uppercase letters
+    - The cell has a lighter background-color
+    - The cell has a muted text color
+    - The cell has a different font style (italic)
+
+#### Scenario 1b: The user identifies a NULL value in a record preview
+
+#### Steps for 1b
+
+1. The user edits a value in a column that links to records from another table
+2. The user sees the list of records displayed in the record selection dropdown
+3. The user identifies a record that contains a NULL value
+    - The field contains the word NULL in all uppercase letters
+    - The text has a muted text color
+    - The text has a different font style (italic)
+
+### Scenario 2: User edits a NULL Value
+
+#### Scenario 2a: The data type is text
+
+#### Steps for 2a
+
+1. The user edits a NULL value in a column that is set to text data type
+2. The user double-clicks the value cell and enters edit mode
+3. The user enters a value and leaves edit mode by hitting enter or clicking outside of the active cell
+4. The NULL value is replaced by the new value
+
+#### Scenario 2b: The data type is text, and the value is empty
+
+#### Steps for 2b
+
+1. The user edits a NULL value in a column that is set to text data type
+2. The user double-clicks the value cell and enters edit mode
+3. The user doesn't enter a value and leaves edit mode by hitting enter or clicking outside of the active cell
```

Comment:
I know we talked about this already in the discussion, but now that I see it in action in the video, I'm not sure about it. Sometimes, when I'm working with spreadsheets or tables, I just select a cell to highlight it for myself to help me think. In this case, if a user highlights the cell and stops highlighting it, then the value of that cell in the database will change, which I don't think users intend to do.

Instead, I think that we should only set the cell to an empty string if the user types something and then deletes it, not just if they enter edit mode and leave it. If they enter edit mode and leave it, it should stay NULL.

Reply:
Is there a differentiation between "selecting" a cell, and "edit mode" on a cell?  For me, that's the most natural, and then this would be less of an issue.  @ghislaineguerin mentions "double-clicking" to enter edit mode, which is similar to how spreadsheets work (at least to my recollection).  If that's the case, it might make sense to somehow make the fact that you've entered "edit mode" very obvious and intentional.  For example, with LibreOffice, the cell highlight changes color and there's also an edit box in another place that activates when you're in "edit mode", but when you're just cursoring around, those things don't happen.  Just an idea.

I agree that we should try to make it an intentional act to set the cell to an empty string, though.  It's more normal to leave things as null than empty strings for most DB schemata.

Reply:
> we should only set the cell to an empty string if the user types something and then deletes it, not just if they enter edit mode and leave it. If they enter edit mode and leave it, it should stay NULL

I agree. 

I think it will be rare for the user to intentionally change a cell value from NULL to empty string. However it will be more common for users to begin editing a cell (perhaps on accident) then stop editing that cell. We should avoid assigning a rarely desired outcome to a more commonly performed action.

**Further**, I think that when the user enters edit mode for a cell with a null value, the cell should _look different_ that it would with an empty string, even in edit mode. I think we should still display "NULL" (in italic) and place the user's cursor at the start of the text. Then when the user begins to type, the "NULL" text will disappear.

Converting a NULL value to an empty string would look like this:

1. Observe NULL in gray italic.
2. Double click.
3. Cell indicates the user is in edit mode by displaying thick border.
4. Cell indicates its value is NULL by displaying NULL in gray italic.
5. Cell places the user's cursor furthest to the left, before the NULL indicator text.
6. User enters any character, e.g. "a".
7. Cell displays "a". The NULL text disappears.
8. User presses `Backspace`.
9. Cell displays as empty -- no characters, no NULL indicator.
10. User presses `Enter` or blurs focus.
11. Cell exits edit mode and the value is saved as an empty string.

Reply:
I think everyone have covered this concern very well. I am on agreement with @seancolsen's suggestion to show null even during edit mode when user has not begun to edit yet.

Reply:
Not sure if this is too late in this conversation or appropriate, but I wonder if you could allow the user to enter edit mode and just hit `backspace` rather than have to create "fake text" and delete it in order to get an empty string.